pandas/_libs/tslib.pyx

```diff
@@ -329,9 +344,13 @@ def array_with_unit_to_datetime(ndarray values, object unit,
 
     if unit == 'ns':
         if issubclass(values.dtype.type, np.integer):
-            return values.astype('M8[ns]'), tz
-        # This will return a tz
-        return array_to_datetime(values.astype(object), errors=errors)
+            result = values.astype('M8[ns]')
+        else:
+            result, tz = array_to_datetime(values.astype(object), errors=errors)
+        if mask is not None:
+            iresult = result.view('i8')
+            iresult[mask] = NPY_NAT
```

Comment:
This is going to mutate `result` inplace. Is that what we want?

Given the subtleties here, does it make sense for the *calling* function to handle this, rather than passing a mask down.

Reply:
yes i think this is worth it, the calling function is aleady way too complex, and we *already* deal with mask here. the mutating here is a common pattern (we mutate on the i8 and auctally return the m8)